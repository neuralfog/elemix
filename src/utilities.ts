import { activeRenderers } from './renderers';

export const ref = <Value>(
    value: Value | undefined = undefined,
): { value?: Value } => {
    return {
        value,
    };
};

export const fastUID = (): string =>
    Math.floor(performance.now() * 1000).toString(36) +
    Math.random().toString(36).slice(2, 6);

export function render(): Promise<boolean> {
    return new Promise((resolve) => {
        const check = () => {
            if (activeRenderers.size === 0) {
                resolve(false);
            } else {
                setTimeout(check, 0);
            }
        };
        check();
    });
}

export const camelToKebabCase = (input: string): string => {
    return (
        input.match(
            /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|[0-9]*$)|[A-Z]?[a-z]+|[A-Z]|[0-9]+/g,
        ) || []
    )
        .map((x) => x.toLowerCase())
        .join('-');
};
