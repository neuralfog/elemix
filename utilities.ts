export { mergeClasses } from '@neuralfog/elemix-renderer/utils';
export * from './src/utilities';
