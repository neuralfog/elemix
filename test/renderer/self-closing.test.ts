import { expect, test, describe, beforeEach } from 'vitest';
import { html } from '@neuralfog/elemix-renderer';
import { Present } from '@neuralfog/elemix-testing';
import { render } from '../../utilities';
import { HTML } from '@neuralfog/elemix-testing/snapshots';

import '../fixtures/renderer/SelfClosingTags';
import {
    type SelfClosingTags,
    State,
} from '../fixtures/renderer/SelfClosingTags';

describe('Renderer Self Closing Tags', () => {
    beforeEach(() => {
        document.body.innerHTML = '';
    });

    test('single', async () => {
        const presenter = new Present().screen(html`<self-closing-tags />`);
        await render();
        expect(HTML(presenter.root())).toMatchSnapshot();
    });

    test('start', async () => {
        const presenter = new Present().screen(html`<self-closing-tags />`);
        const component =
            presenter.getComponent<SelfClosingTags>('self-closing-tags');

        component.state.case = State.START;
        await render();
        expect(HTML(presenter.root())).toMatchSnapshot();
    });

    test('end', async () => {
        const presenter = new Present().screen(html`<self-closing-tags />`);
        const component =
            presenter.getComponent<SelfClosingTags>('self-closing-tags');

        component.state.case = State.END;
        await render();
        expect(HTML(presenter.root())).toMatchSnapshot();
    });

    test('middle', async () => {
        const presenter = new Present().screen(html`<self-closing-tags />`);
        const component =
            presenter.getComponent<SelfClosingTags>('self-closing-tags');

        component.state.case = State.MIDDLE;
        await render();
        expect(HTML(presenter.root())).toMatchSnapshot();
    });
});
