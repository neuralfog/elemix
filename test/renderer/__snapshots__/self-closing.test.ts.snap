// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renderer Self Closing Tags > end 1`] = `
"
<self-closing-tags><!-- #shadowroot -->
  <p>hello there</p>
  <self-closed><!-- #shadowroot -->
    <div>Self Closed!!</div>
  </self-closed>
</self-closing-tags>
"
`;

exports[`Renderer Self Closing Tags > middle 1`] = `
"
<self-closing-tags><!-- #shadowroot -->
  <p>hello there</p>
  <self-closed><!-- #shadowroot -->
    <div>Self Closed!!</div>
  </self-closed>
  <p>hello there</p>
</self-closing-tags>
"
`;

exports[`Renderer Self Closing Tags > single 1`] = `
"
<self-closing-tags><!-- #shadowroot -->
  <self-closed><!-- #shadowroot -->
    <div>Self Closed!!</div>
  </self-closed>
</self-closing-tags>
"
`;

exports[`Renderer Self Closing Tags > start 1`] = `
"
<self-closing-tags><!-- #shadowroot -->
  <self-closed><!-- #shadowroot -->
    <div>Self Closed!!</div>
  </self-closed>
  <p>Hello There</p>
</self-closing-tags>
"
`;
